server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,478 @@
+                : "runNowOrBacklog must be invoked before actually running the merge task";
+            try {
+                beforeMerge(onGoingMerge);
+                try {
+                    if (mergeStartTimeNS.compareAndSet(0L, System.nanoTime()) == false) {
+                        throw new IllegalStateException("The merge task is already started");
+                    }
+                    mergeTracking.mergeStarted(onGoingMerge);
+                    if (verbose()) {
+                        message(String.format(Locale.ROOT, "merge task %s start", this));
+                    }
+                    try {
+                        doMerge(mergeSource, onGoingMerge.getMerge());
+                        if (verbose()) {
+                            message(
+                                String.format(
+                                    Locale.ROOT,
+                                    "merge task %s merge segment [%s] done estSize=%.1f MB (written=%.1f MB) "
+                                        + "runTime=%.1fs (stopped=%.1fs, paused=%.1fs) rate=%s",
+                                    this,
+                                    getSegmentName(onGoingMerge.getMerge()),
+                                    bytesToMB(onGoingMerge.getMerge().estimatedMergeBytes),
+                                    bytesToMB(rateLimiter.getTotalBytesWritten()),
+                                    nsToSec(System.nanoTime() - mergeStartTimeNS.get()),
+                                    nsToSec(rateLimiter.getTotalStoppedNS()),
+                                    nsToSec(rateLimiter.getTotalPausedNS()),
+                                    rateToString(rateLimiter.getMBPerSec())
+                                )
+                            );
+                        }
+                    } finally {
+                        long tookMS = TimeValue.nsecToMSec(System.nanoTime() - mergeStartTimeNS.get());
+                        mergeTracking.mergeFinished(onGoingMerge.getMerge(), onGoingMerge, tookMS);
+                    }
+                } finally {
+                    if (verbose()) {
+                        message(String.format(Locale.ROOT, "merge task %s end", this));
+                    }
+                    afterMerge(onGoingMerge);
+                }
+            } finally {
+                if (verbose()) {
+                    message(String.format(Locale.ROOT, "merge task %s end", this));
+                }
+                mergeDone(this);
+                try {
+                    // kick-off next merge, if any
+                    merge(mergeSource, MergeTrigger.MERGE_FINISHED);
+                } catch (@SuppressWarnings("unused") AlreadyClosedException ace) {
+                    // OK, this is what the {@code ConcurrentMergeScheduler} does
+                }
+            }
+        }
+
+        void abortOnGoingMerge() {
+            // This interrupts the IndexWriter.
```

Comment:
I did not see `setAborted` "interrupting" as in java interrupt, I am not sure exactly what you mean here? Perhaps use another word than interrupt since it has a special meaning (unless I got it wrong ofc).

Reply:
I'll rephrase it. The `IndexWriter` checks the `aborted` flag before various steps, and reverts the merge if set.

Reply:
[0959714](https://github.com/elastic/elasticsearch/pull/120869/commits/0959714810e8721a1d117e8d4434e05659d0eb32)